py/dml/ctree.py

```diff
@@ -3162,6 +3167,9 @@ def __init__(self, site, node, trait, indices,
         if not isinstance(indices, tuple):
             raise ICE(site, 'bad indices: %r' % (indices,))
         assert len(indices) == node.dimensions
+        for index in indices:
+            if isinstance(index, NonValue):
+                raise index.exc()
```

Comment:
Turn this into an assert, instead. This is caught in `__init__`, and we should not emit exceptions this late.

Reply:
ignoring after IM discussion

Reply:
Sorry, read the diffs wrong. I thought this addition was under Dereference.read(). Ignore!